### DISCUSSION TITLE
Can't install any ARM version of python

### DISCUSSION THREAD

**Discussion by russ-mcbride** [pyenv/pyenv#2878] [General]
I'm using homebrew tcl-tk, readline, openssl, zlib, and gcc. 

pyenv: 2.3.35
Mac OS: 14.2.1
Homebrew: 4.22

CC=gcc pyenv install 3.10.13
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.10.13.tar.xz...
-> https://www.python.org/ftp/python/3.10.13/Python-3.10.13.tar.xz
Installing Python-3.10.13...
python-build: use tcl-tk from homebrew
python-build: use readline from homebrew
python-build: use zlib from homebrew
...
raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/private/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/python-build.20240101224952.75879/Python-3.10.13/python.exe', '-W', 'ignore::DeprecationWarning', '-c', '\nimport runpy\nimport sys\nsys.path = [\'/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/tmpk0lwgq2z/setuptools-65.5.0-py3-none-any.whl\', \'/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/tmpk0lwgq2z/pip-23.0.1-py3-none-any.whl\'] + sys.path\nsys.argv[1:] = [\'install\', \'--no-cache-dir\', \'--no-index\', \'--find-links\', \'/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/tmpk0lwgq2z\', \'--root\', \'/\', \'--upgrade\', \'setuptools\', \'pip\']\nrunpy.run_module("pip", run_name="__main__", alter_sys=True)\n']' died with <Signals.SIGABRT: 6>.
make: *** [install] Error 1

Suggestions?

**Comment by native-api** [comment]
https://github.com/search?q=repo%3Apyenv%2Fpyenv+died+with+%3CSignals.SIGABRT%3A+6&type=issues , in particular, https://github.com/pyenv/pyenv/issues/2707 and https://github.com/pyenv/pyenv/issues/2815#issuecomment-1753294719

**Comment by russ-mcbride** [comment]
Thanks for the reply. I went through those carefully and went through the "Common Build Problems" wiki, in addition to reinstalling Xcode, Homebrew, pyenv, and all sorts of compiler and linker flag settings. Forgive my ignorance as I'm probably screwing up those settings in some way. I can't tell if I've made forward progress or reverse progress but I'm now getting these common two errors but can't seem to solve them after 3 days of working on this:

```
WARNING: The Python readline extension was not compiled. Missing the GNU readline lib?
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?
```
Is the "Python ssl extension" separate from the one I'm referring to in my install command? I'm confused.
I'm also confused about why setting the flags in my install command gives me different results when those flags are already set in my .zshrc file.


Install command:
```
LDFLAGS="-L/opt/homebrew/opt/openssl@1.1" \
CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1" \
pyenv install -v 3.11.7
```

The initial response from this is:
```
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/python-build.20240103094955.95458 /opt/homebrew/Cellar/openssl@1.1/1.1.1w/lib
Downloading Python-3.11.7.tar.xz...
-> https://www.python.org/ftp/python/3.11.7/Python-3.11.7.tar.xz
/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/python-build.20240103094955.95458/Python-3.11.7 /var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/python-build.20240103094955.95458 /opt/homebrew/Cellar/openssl@1.1/1.1.1w/lib
Installing Python-3.11.7...
python-build: use tcl-tk from homebrew
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
```

So, it obviously sees those libraries.

My environment looks like this:
```
% env
TERM_PROGRAM=Apple_Terminal
SHELL=/bin/zsh
TERM=xterm-256color
TMPDIR=/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/
TERM_PROGRAM_VERSION=452
TERM_SESSION_ID=C6EB3B82-E4BE-4F5C-A738-A6B50375CFC6
USER=russ
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.gJihUjYw6p/Listeners
PATH=/Users/russ/.pyenv/bin:/opt/homebrew:/opt/homebrew/opt/readine:/opt/homebrew/opt/openssl@1.1:/Users/russ/.pyenv/bin:/opt/homebrew:/opt/homebrew/opt/readine:/opt/homebrew/opt/openssl@1.1:/Users/russ/.pyenv/bin:/opt/homebrew:/opt/homebrew/opt/readine:/opt/homebrew/opt/openssl@1.1:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin
LaunchInstanceID=1FF9E430-720A-4401-A85E-45CC8250E050
__CFBundleIdentifier=com.apple.Terminal
PWD=/opt/homebrew/Cellar/openssl@1.1/1.1.1w/lib
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
SHLVL=1
HOME=/Users/russ
LOGNAME=russ
SECURITYSESSIONID=18790
OLDPWD=/opt/homebrew/Cellar/openssl@1.1/1.1.1w/lib
HOMEBREW_PREFIX=/opt/homebrew
HOMEBREW_CELLAR=/opt/homebrew/Cellar
HOMEBREW_REPOSITORY=/opt/homebrew
MANPATH=/opt/homebrew/share/man::
INFOPATH=/opt/homebrew/share/info:
PYENV_ROOT=/Users/russ/.pyenv
LDFLAGS=-L/opt/homebrew/opt/readline/lib:-L/opt/homebrew/lib
LDFFLAGS=-L/opt/homebrew/opt/openssl@1.1/lib:-L/opt/homebrew/lib
CPPFLAGS=-I/opt/homebrew/opt/readline/include:-I/opt/homebrew/opt/openssl@1.1/include
PKG_CONFIG_PATH=/opt/homebrew/opt/readline/lib/pkgconfig
PYTHON_BUILD_HOMEBREW_OPENSSL_FORMULA=openssl@1.1
LANG=en_US.UTF-8
_=/usr/bin/env
```

And my .zshrc file looks like this:
```
setopt no_nomatch

echo "Now reading .zshrc..."

alias arm="env /usr/bin/arch -arm64 /bin/zsh --login"
alias intel="env /usr/bin/arch -x86_64 /bin/zsh --login"

alias brew86="arch -x86_64 /usr/local/bin/brew"
alias brew="arch -arm64 /opt/homebrew/bin/brew"

alias pyenv86="arch -x86_64 pyenv"
alias pyenv="arch -arm64 pyenv"

export LDFLAGS="-L/opt/homebrew/lib"
export LDFFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib:$LDFLAGS"
export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"
export LDFLAGS="-L/opt/homebrew/opt/readline/lib:$LDFLAGS"
export CPPFLAGS="-I/opt/homebrew/opt/readline/include:$CPPFLAGS"
export PKG_CONFIG_PATH="/opt/homebrew/opt/readline/lib/pkgconfig"
export DYLD_LIBRARY_PATH="/opt/homebrew/Cellar/openssl@1.1/1.1.1w/lib:$DYLD_LIBRARY_PATH"
export PYTHON_BUILD_HOMEBREW_OPENSSL_FORMULA="openssl@1.1"

#eval "$(/opt/homebrew/bin/brew shellenv)"

#NOTE: no quotes here for zsh shell
export PATH=/opt/homebrew/opt/openssl@1.1:$PATH
export PATH=/opt/homebrew/opt/readine:$PATH
export PATH=/opt/homebrew:$PATH

#For pyenv:
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
```

I shouldn't put the CPPFLAGS and LDFLAGS in my zsh profile, I realize, but I'm grasping at straws at this point.

Thanks.

**Comment by russ-mcbride** [comment]
Solution--

I realized that each library I added to Homebrew would then trigger a "missing module" error during the install process, counter-intuitively. I simply removed all those libraries from Homebrew, which forced the download of them during the install process and the install proceeded smoothly.

The question remains: What is happening between Pyenv and Homebrew to cause this counterintuitive behavior?

**Reply by native-api** [reply]
Most probably, you're building for a different architecture than `brew` on `PATH` is for. So any libraries you install there produce errors on the link step.

**Reply by russ-mcbride** [reply]
Hmm... well, Brew on my path is for arm64. Most (all?) python versions are still x86, I assume, so maybe that was causing a conflict? I *think* in the past I had  installed versions of python in the past that would give me "arm64" as a result when I ran "import platform; platform.machine(); quit()", but maybe I'm hallucinating....

**Comment by trinitronx** [comment]
>     subprocess.CalledProcessError: Command 
>     '['/private/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/python-build.20240101224952.75879/\
>     Python-3.10.13/python.exe',
>                           ^^^

This seems odd to me... 🤔

You say you're building an ARM version of Python, and using Homebrew... and there are instances of `com.apple.*` and `/private/var/folders/*` and other macOS-specific paths in your shell `env` and log output.  So I assume that means you're on macOS.

Yet it's somehow running the `.exe` version of Python as if it were on Windows?

```zsh
export LDFLAGS="-L/opt/homebrew/lib"
export LDFFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib:$LDFLAGS"
          ^^^ Typo here
```

Also, this setting for the `openssl@1.1` `LDFLAGS` has a typo in the variable name:

| Was           | Should be |
|---------------|-----------|
| `LDFFLAGS`    | `LDFLAGS` |

Something might be causing `pyenv` + `python-build` to think you're on Windows and then tries to run `python.exe` with the CLI flags mentioned here:

```console
subprocess.CalledProcessError: Command '['/private/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/python-build.20240101224952.75879/Python-3.10.13/python.exe', '-W', 'ignore::DeprecationWarning', '-c', '\nimport runpy\nimport sys\nsys.path = ['/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/tmpk0lwgq2z/setuptools-65.5.0-py3-none-any.whl', '/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/tmpk0lwgq2z/pip-23.0.1-py3-none-any.whl'] + sys.path\nsys.argv[1:] = ['install', '--no-cache-dir', '--no-index', '--find-links', '/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/tmpk0lwgq2z', '--root', '/', '--upgrade', 'setuptools', 'pip']\nrunpy.run_module("pip", run_name="main", alter_sys=True)\n']' died with <Signals.SIGABRT: 6>.
```

So, it looks like this is `python-build` running:

```zsh
pip install --no-cache-dir --no-index --find-links '/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/tmpk0lwgq2z' --root / --upgrade setuptools
```

However, it's trying to run this using `python.exe`[^1] and dies with `SIGABRT`. It could be due to an exec format error if the `python.exe` binary is not the correct executable format expected on ARM64 macOS.

Another thing to check is whether you're on an Apple Silicon Mac and if `Terminal.app` is running under Rosetta 2 or native as `arm64` / `aarch64`.  You can detect this by running the following inside your shell session:

```zsh
## Figure out OSX version (source: https://www.opscode.com/chef/install.sh)
function detect_platform_version() {
  # Matching the tab-space with sed is error-prone
  platform_version=$(sw_vers | awk '/^ProductVersion:/ { print $2 }')

  major_version=$(echo "$platform_version" | cut -d. -f1,2)

  # x86_64 Apple hardware often runs 32-bit kernels (see OHAI-63)
  # macOS Monterey + Apple M1 Silicon (arm64) gives empty string for this x86_64 check
  x86_64=$(sysctl -n hw.optional.x86_64)
  arm64=$(sysctl -n hw.optional.arm64)
  if [[ "$x86_64" == '1' ]]; then
    machine="x86_64"
  elif [[ "$arm64" == '1' ]]; then
    machine="arm64"
  fi

  echo "Detected:" >&2
  echo "platform_version: ${platform_version}" >&2
  echo "major_version: ${major_version}" >&2
  echo "machine: ${machine}" >&2
}

detect_platform_version
```

This should print "`machine: arm64`" if it's natively running as ARM64 on Apple Silicon, and "`machine: x86_64`" if it's running under Intel `x86_64` emulation under Rosetta 2.

Run the `CC=gcc pyenv install 3.10.13` command again and copy the file path to the `python.exe` that it mentions in the error.  Then check which executable format it is (according to `libmagic` file header rules) with the `file` command.  For example:

```zsh
file /private/var/folders/xk/vp0pdf492tl589vtnykls31r0000gn/T/python-build.20240101224952.75879/Python-3.10.13/python.exe
```

This should show you that it's either an `arm64` Mach-O binary, or a "[Universal binary][1]", a.k.a a "fat" Mach-O binary bundle[^2].

If it shows up as a Windows executable, then something's wrong with your OS being detected properly.

One final thing I noticed is that you're using `gcc` rather than the default `clang`.  Typically, Homebrew and XCode on macOS use the `clang` toolchain.  Generally, it's best to stick to the default toolchain used by Homebrew.  Although technically, if you know what you're doing, you could use `gcc` from Homebrew to compile one-off things.

So, remove and unset the `CC` variable that you had here: `CC=gcc pyenv install 3.10.13`

```zsh
unset CC

pyenv install 3.10.13
```

[^1]: Strangely, it has an `.exe` file extension as typically found on Windows.  Technically, the underlying file content could be any type, but maybe `python-build` defaults to using an `.exe` file extension?  Either way, it looks strange to me.  The `pip` command is really what is being run to install `setuptools`.  That is running through `pip`'s python module mode of execution, like `python -m pip [...args...]`
[^2]: [Apple docs show how to build these "fat" Universal binaries with the colorfully-named "`lipo`" tool here][2].  [Here is a deep-dive into what is inside these binary formats][3].

[1]: https://en.wikipedia.org/wiki/Universal_binary
[2]: https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary
[3]: https://www.jviotti.com/2021/07/23/a-deep-dive-on-macos-universal-binaries.html
